### ISSUE TITLE
Ansible tries to evaluate registered state

### ISSUE THREAD

**Issue by dene14** [ansible/ansible#21705]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
Core

##### ANSIBLE VERSION
```
ansible 2.2.1.0
  config file = 
  configured module search path = Default w/o overrides
```

##### CONFIGURATION
Default

##### OS / ENVIRONMENT
N/a

##### SUMMARY
Ansible tries to evaluate registered state variable. If you register state of git repo checkout and there is something like "{{something}}" in git diff, ansible will fail most probably in a step where you try to check "state.changed" key because "{{something}}" hasn't been found in ansible variables stack.

##### STEPS TO REPRODUCE
```
- name: Pull source from the repository
  become_user: "{{ app_user }}"
  git: repo={{ repository }}
       dest={{ project_root }}
       version={{ repository_branch }}
       key_file={{ repository_dest_ssh_private_key_file }}
       accept_hostkey=yes
       force=yes
       update=yes
       track_submodules=no
  register: repo_state
  when: "environment_name != 'vagrant'"
#  notify: restart supervisor apps

- name: Set repo change flag
  set_fact:
     repo_state_changed: "{{ repo_state.changed }}"

- name: Notify supervisor on repo change if applicable
  shell: echo "repo state changed"
#  changed_when: repo_state.changed
  notify: restart supervisor apps
  when: "environment_name != 'vagrant' and repo_state_changed and app_name != 'app-' + project_name + '-web'"
```
This example looks ugly however it represents all 4 solutions that I've tried:
- notify supervisor directly from repo checkout step
- use changed_when in a separate task (seems #21298)
- use dumb task to check state conditionally in `when` clause
- use set_fact prior to dumb notification task

none of them work, let me know of there any other workaround for this.

##### EXPECTED RESULTS
I expect that "repo_state.changed" is the only key processed by ansible and returned to evaluation, not the whole state's dict. I cannot even imagine any case where you need to evaluate git's diff.

##### ACTUAL RESULTS
```
fatal: [testbox.local]: FAILED! => {"failed": true, "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: {u'after': u'f5738c02baa2db7e09cbb217a69ad36039627c02', u'warnings': [], u'changed': True, u'diff': {u'prepared': u'diff --git a/src/profile/profile.jade b/src/profile/profile.jade\\nindex dfc3184..e228eef 100644\\n--- a/src/profile/profile.jade\\n+++ b/src/profile/profile.jade\\n@@ -3,5 +3,5 @@ div(am-width=\"x6\" am-bg=\"light-grey\" am-height=\"fit\")\\n   div(am-height=\"x2\" am-gutter am-Grid-Row=\"middle\" am-border=\"dotted\")\\n     div(am-Button=\"fit\" ng-click=\"profileCtrl.logout()\") Logout\\n \\n-  div(am-height=\"x2\" am-gutter am-Grid-Row=\"middle\" am-border=\"dotted\")\\n+  //div(am-height=\"x2\" am-gutter am-Grid-Row=\"middle\" am-border=\"dotted\")\\n     a(am-Button=\"fit\" href=\"{{profileCtrl.resetPasswordLink}}\") Reset Password\\n'}, u'before': u'3ddd346b76f5814e3c2213083797e1d67eaa7c82'}: 'profileCtrl' is undefined\n\nThe error appears to have been in '/var/lib/jenkins/workspace/mimosa-jobs/mimosa-devel-deploy/roles/app-roles/sys-init/tasks/prep_repo.yml': line 28, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Set repo change flag\n  ^ here\n"}

msg: the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: {u'after': u'f5738c02baa2db7e09cbb217a69ad36039627c02', u'warnings': [], u'changed': True, u'diff': {u'prepared': u'diff --git a/src/profile/profile.jade b/src/profile/profile.jade
index dfc3184..e228eef 100644
--- a/src/profile/profile.jade
+++ b/src/profile/profile.jade
@@ -3,5 +3,5 @@ div(am-width="x6" am-bg="light-grey" am-height="fit")
   div(am-height="x2" am-gutter am-Grid-Row="middle" am-border="dotted")
     div(am-Button="fit" ng-click="profileCtrl.logout()") Logout
 
-  div(am-height="x2" am-gutter am-Grid-Row="middle" am-border="dotted")
+  //div(am-height="x2" am-gutter am-Grid-Row="middle" am-border="dotted")
     a(am-Button="fit" href="{{profileCtrl.resetPasswordLink}}") Reset Password
'}, u'before': u'3ddd346b76f5814e3c2213083797e1d67eaa7c82'}: 'profileCtrl' is undefined
```

**Comment by evgeni** [comment]
This is [CVE-2017-7466](https://bugzilla.redhat.com/show_bug.cgi?id=CVE-2017-7466)

Fixed in 2.2: https://github.com/ansible/ansible/commit/7ff9fa52cfcef2065f0db80d85dd94b9b754839c
Fixed in 2.3/devel: https://github.com/ansible/ansible/commit/0d418789a298561fded9bce977d34babc9097079

**Comment by sivel** [comment]
This issue has been resolved in the v2.3 release.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
